src/System.Runtime.InteropServices/tests/System/Runtime/InteropServices/Marshal/ReadWrite/ObjectTests.cs

```diff
@@ -0,0 +1,106 @@
+
+            object boxedBlittableStruct = (object)blittableStruct;
+
+            int offsetOfB = Marshal.OffsetOf<BlittableStruct>("_b").ToInt32();
+            int offsetOfC = Marshal.OffsetOf<BlittableStruct>("_c").ToInt32();
+            int offsetOfD = Marshal.OffsetOf<BlittableStruct>("_d").ToInt32();
+            int offsetOfP = Marshal.OffsetOf<BlittableStruct>("_p").ToInt32();
+
+            Assert.Equal(Marshal.ReadInt32(boxedBlittableStruct, 0), 200);
+            Assert.Equal(Marshal.ReadInt32(boxedBlittableStruct, offsetOfB), 300);
+            Assert.Equal(Marshal.ReadByte(boxedBlittableStruct, offsetOfC), 10);
+            Assert.Equal(Marshal.ReadInt16(boxedBlittableStruct, offsetOfD), 123);
+            Assert.Equal(Marshal.ReadIntPtr(boxedBlittableStruct, offsetOfP), new IntPtr(100));
+
+            Marshal.WriteInt32(boxedBlittableStruct, 0, 300);
+            Marshal.WriteInt32(boxedBlittableStruct, offsetOfB, 400);
+            Marshal.WriteByte(boxedBlittableStruct, offsetOfC, 20);
+            Marshal.WriteInt16(boxedBlittableStruct, offsetOfD, 144);
+
+            Marshal.WriteIntPtr(boxedBlittableStruct, offsetOfP, new IntPtr(500));
+
+            Assert.Equal(((BlittableStruct)boxedBlittableStruct)._a, 300);
+            Assert.Equal(((BlittableStruct)boxedBlittableStruct)._b, 400);
+            Assert.Equal(((BlittableStruct)boxedBlittableStruct)._c, 20);
+            Assert.Equal(((BlittableStruct)boxedBlittableStruct)._d, 144);
+            Assert.Equal(((BlittableStruct)boxedBlittableStruct)._p, new IntPtr(500));
+        }
+
+        [Fact]
+        public void TestStructWithReferenceType()
+        {
+            StructWithReferenceTypes structWithReferenceTypes = new StructWithReferenceTypes();
+            structWithReferenceTypes._ptr = new IntPtr(100);
+            structWithReferenceTypes._str = "ABC";
+            structWithReferenceTypes._byValArr = new int[10] { 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 };
+
+            object boxedStruct = (object)structWithReferenceTypes;
+
+            int offsetOfStr = Marshal.OffsetOf<StructWithReferenceTypes>("_str").ToInt32();
+            int offsetOfByValArr = Marshal.OffsetOf<StructWithReferenceTypes>("_byValArr").ToInt32();
+
+            Assert.Equal(Marshal.ReadInt32(boxedStruct, 0), 100);
+            Assert.NotEqual(Marshal.ReadIntPtr(boxedStruct, offsetOfStr), IntPtr.Zero);
+            Assert.Equal(Marshal.ReadInt32(boxedStruct, offsetOfByValArr + sizeof(int) * 2), 3);
+
+            Marshal.WriteInt32(boxedStruct, 0, 200);
+            Marshal.WriteInt32(boxedStruct, offsetOfByValArr + sizeof(int) * 9, 100);
+
+            Assert.Equal(((StructWithReferenceTypes)boxedStruct)._ptr, new IntPtr(200));
+            Assert.Equal(((StructWithReferenceTypes)boxedStruct)._byValArr[9], 100);
+            Assert.Equal(((StructWithReferenceTypes)boxedStruct)._byValArr[8], 9);
+            Assert.Equal(((StructWithReferenceTypes)boxedStruct)._str, "ABC");
```

Comment:
All of these Assert.Equal should have their arguments swapped.  The first argument to Equal is the expected value, the second argument is the actual value.

Reply:
Fixed, see in next commit